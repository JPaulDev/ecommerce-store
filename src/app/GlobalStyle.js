import { createGlobalStyle } from 'styled-components';
import openSansRegular from '../assets/fonts/OpenSans-Regular.woff';
import openSansMedium from '../assets/fonts/OpenSans-Medium.ttf';
import openSansSemiBold from '../assets/fonts/OpenSans-SemiBold.woff';
import openSansBold from '../assets/fonts/OpenSans-Bold.woff';
import uniSansBook from '../assets/fonts/UniSans-Book.woff2';
import uniSansRegular from '../assets/fonts/UniSans-Regular.woff2';
import uniSansSemiBold from '../assets/fonts/UniSans-SemiBold.woff2';
import uniSansBold from '../assets/fonts/UniSans-Bold.woff';

const GlobalStyle = createGlobalStyle`
  :root {
    --color-orange-1: #ff9110;
    --color-grey-1: #434343;
    --color-grey-2: #333;
    --color-grey-3: #b3b3b3;
    --color-grey-4: #2f2f2f;
    --color-grey-5: #7e7e7e;
    --color-grey-6: #f6f6f6;
    --color-grey-7: #9e9e9e;
    --color-grey-8: #5b5b5b;
    --color-grey-9: #e1e1e1;
    --color-grey-10: #bcbec0;
    --color-grey-11: #cfcfcf;
    --color-grey-12: #c7c7c7;
    --color-blue-1: #3d70b7;
    --color-blue-2: #1766a6;
  }

  * {
    font-family: 'Open Sans', -apple-system, BlinkMacSystemFont, 'Segoe UI',
      'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
      'Helvetica Neue', sans-serif;
    box-sizing: border-box;
    margin: 0;
    padding: 0;
    border: 0;
  }

  html {
    font-size: 16px;
  }

  #root {
    position: relative;
    min-height: 100vh;
    overflow: clip;
  }

  @font-face {
    font-family: 'Open Sans';
    src: url(${openSansRegular}) format('woff');
    font-weight: 400;
  }

  @font-face {
    font-family: 'Open Sans';
    src: url(${openSansMedium}) format('truetype');
    font-weight: 500;
  }

  @font-face {
    font-family: 'Open Sans';
    src: url(${openSansSemiBold}) format('woff');
    font-weight: 600;
  }

  @font-face {
    font-family: 'Open Sans';
    src: url(${openSansBold}) format('woff');
    font-weight: 700;
  }

  @font-face {
    font-family: 'Uni Sans Book';
    src: url(${uniSansBook}) format('woff2');
  }

  @font-face {
    font-family: 'Uni Sans';
    src: url(${uniSansRegular}) format('woff2');
    font-weight: 400;
  }

  @font-face {
    font-family: 'Uni Sans';
    src: url(${uniSansSemiBold}) format('woff2');
    font-weight: 600;
  }

  @font-face {
    font-family: 'Uni Sans';
    src: url(${uniSansBold}) format('woff');
    font-weight: 700;
  }
`;

export default GlobalStyle;
