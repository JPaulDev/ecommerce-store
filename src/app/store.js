import { combineReducers, configureStore } from '@reduxjs/toolkit';
import basketReducer from '../features/basket/basketSlice';

const rootReducer = combineReducers({
  basket: basketReducer,
});

export const setupStore = (preloadedState) =>
  configureStore({
    reducer: rootReducer,
    preloadedState,
  });
