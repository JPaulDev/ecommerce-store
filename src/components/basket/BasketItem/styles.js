import styled from 'styled-components';

export const ListItem = styled.li`
  align-items: center;
  border-bottom: 1px solid var(--color-grey-9);
  display: flex;
  padding: 10px 0 30px;

  img {
    height: auto;
    margin-right: 20px;
    width: 120px;

    @media screen and (min-width: 900px) {
      width: 140px;
      height: auto;
    }
  }
`;

export const ProductDescription = styled.p`
  border-bottom: 3px solid var(--color-grey-9);
  font-family: 'Uni Sans', sans-serif;
  font-weight: 600;
  line-height: 1.4;
  margin: 3px 0 5px;
  padding-bottom: 10px;
`;

export const Container = styled.div`
  align-items: center;
  display: flex;
`;

export const RemoveProduct = styled.button`
  background: none;
  border-left: 1px solid var(--color-grey-10);
  cursor: pointer;
  font-size: 0.9rem;
  margin-left: 20px;
  padding-left: 20px;

  &:hover {
    text-decoration: underline;
  }
`;

export const QuantitySelector = styled.div`
  background-color: var(--color-grey-6);
  border-radius: 5px;
  display: inline-flex;
  justify-content: space-between;
  margin-top: 10px;
  min-width: 105px;
  padding: 6px;
`;

export const IncreaseButton = styled.button`
  background-image: linear-gradient(to top, #3b7739, #55a952);
  border-radius: 4px;
  box-shadow: var(--shadow-button);
  color: white;
  cursor: pointer;
  display: grid;
  height: 25px;
  place-items: center;
  width: 25px;

  &:hover {
    background-image: linear-gradient(to top, #326830, #4a9b47);
    box-shadow: var(--shadow-button-hover);
  }
`;

export const DecreaseButton = styled(IncreaseButton)`
  background-image: linear-gradient(to top, #993b39, #cc5552);

  &:hover {
    background-image: linear-gradient(to top, #901c30, #cf3747);
  }
`;

export const Quantity = styled.span`
  align-items: center;
  color: #0c9444;
  display: flex;
  font-weight: 600;
  justify-content: center;
  min-width: 35px;
  padding: 0 5px;
`;
