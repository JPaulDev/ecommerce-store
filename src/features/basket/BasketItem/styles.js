import styled, { css } from 'styled-components';

const flexMixin = css`
  align-items: center;
  display: flex;
`;

export const Container = styled.div`
  ${flexMixin}
`;

export const ListItem = styled.li`
  border-bottom: 1px solid var(--color-grey-12);
  padding: 10px 0 30px;
  ${flexMixin}

  img {
    height: auto;
    margin-right: 20px;
    width: 120px;

    @media screen and (min-width: 900px) {
      width: 140px;
      height: auto;
    }
  }
`;

export const ProductDescription = styled.p`
  border-bottom: 3px solid var(--color-grey-12);
  font-family: 'Uni Sans', sans-serif;
  font-weight: var(--font-weight-semibold);
  line-height: 1.5;
  margin: 3px 0 5px;
  padding-bottom: 10px;
`;

export const RemoveProduct = styled.button`
  border-left: 1px solid var(--color-grey-9);
  cursor: pointer;
  font-size: var(--font-size-14);
  margin-left: 20px;
  padding-left: 20px;

  &:hover {
    text-decoration: underline;
  }
`;

export const QuantitySelector = styled.div`
  background-color: var(--color-grey-15);
  border-radius: 5px;
  display: inline-flex;
  justify-content: space-between;
  margin-top: 10px;
  min-width: 105px;
  padding: 6px;
`;

export const IncreaseButton = styled.button`
  background-image: linear-gradient(
    to top,
    var(--color-green-2),
    var(--color-green-5)
  );
  border-radius: 4px;
  box-shadow: var(--box-shadow-md);
  color: white;
  cursor: pointer;
  display: grid;
  height: 25px;
  place-items: center;
  width: 25px;

  &:hover {
    background-image: linear-gradient(
      to top,
      var(--color-green-1),
      var(--color-green-4)
    );
    box-shadow: var(--box-shadow-hover);
  }
`;

export const DecreaseButton = styled(IncreaseButton)`
  background-image: linear-gradient(
    to top,
    var(--color-red-2),
    var(--color-red-5)
  );

  &:hover {
    background-image: linear-gradient(
      to top,
      var(--color-red-1),
      var(--color-red-4)
    );
  }
`;

export const Quantity = styled.div`
  align-items: center;
  color: var(--color-green-3);
  display: flex;
  font-weight: var(--font-weight-semibold);
  justify-content: center;
  min-width: 35px;
  padding: 0 5px;
  ${flexMixin}
`;
