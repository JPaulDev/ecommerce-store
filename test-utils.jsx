import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { setupStore } from './src/app/store';
import { UIProvider } from './src/contexts/UIContext';
import theme from './src/styles/theme';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <UIProvider>{children}</UIProvider>
        </Provider>
      </ThemeProvider>
    );
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';
export { renderWithProviders as render };
