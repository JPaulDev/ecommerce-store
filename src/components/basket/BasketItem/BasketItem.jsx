import Image from 'next/future/image';
import { memo } from 'react';
import { useDispatch } from 'react-redux';
import {
  decrementQuantity,
  incrementQuantity,
  removeProduct,
} from '../../../features/basket/basketSlice';
import usePrice from '../../../lib/hooks/usePrice';
import { Plus, Minus } from '../../icons';
import { PartNumber, Price, PreviousPrice, StockIndicator } from '../../common';
import * as Styled from './styles';

function BasketItem({ product }) {
  const dispatch = useDispatch();

  const { price, previousPrice } = usePrice({
    isOnSale: product.isOnSale,
    discount: product.discount * product.quantity,
    price: product.price * product.quantity,
    currencyCode: product.currencyCode,
  });

  const handleIncrementQuantity = () =>
    dispatch(incrementQuantity(product.sku));
  const handleDecrementQuantity = () =>
    dispatch(decrementQuantity(product.sku));
  const handleRemoveProduct = () => dispatch(removeProduct(product.sku));

  return (
    <Styled.ListItem as="li">
      <Image
        src={product.imageUrl}
        width={200}
        height={200}
        alt=""
        quality={85}
        priority="true"
      />
      <div>
        <PartNumber fontSize="var(--font-size-13)" sku={product.sku} />
        <Styled.ProductDescription>
          {product.name}, {product.description}
        </Styled.ProductDescription>
        <Styled.Container>
          <Price
            price={price}
            fontSize="var(--font-size-24)"
            fontWeight="var(--font-weight-medium)"
          />
          <Styled.RemoveProduct type="button" onClick={handleRemoveProduct}>
            Remove Item
          </Styled.RemoveProduct>
        </Styled.Container>
        {product.isOnSale && (
          <PreviousPrice price={previousPrice} fontSize="var(--font-size-14)" />
        )}
        <StockIndicator stockStatus={product.stockStatus} marginTop="5px" />
        <Styled.QuantitySelector>
          <Styled.DecreaseButton
            type="button"
            title="Decrease Quantity"
            aria-label="Decrease Quantity"
            onClick={handleDecrementQuantity}
          >
            <Minus width={15} height={17} />
          </Styled.DecreaseButton>
          <Styled.Quantity data-testid="quantity">
            {product.quantity}
          </Styled.Quantity>
          <Styled.IncreaseButton
            type="button"
            title="Increase Quantity"
            aria-label="Increase Quantity"
            onClick={handleIncrementQuantity}
          >
            <Plus width={15} height={17} />
          </Styled.IncreaseButton>
        </Styled.QuantitySelector>
      </div>
    </Styled.ListItem>
  );
}

export default memo(BasketItem);
