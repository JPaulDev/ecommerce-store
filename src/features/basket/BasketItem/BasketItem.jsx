import Image from 'next/future/image';
import { memo } from 'react';
import { useDispatch } from 'react-redux';
import { Minus, Plus } from '../../../components/icons';
import {
  PartNumber,
  PreviousPrice,
  Price,
  StockDisplay,
} from '../../../components/product';
import { usePrice } from '../../../hooks';
import {
  decrementQuantity,
  incrementQuantity,
  removeProduct,
} from '../basketSlice';
import * as Styled from './styles';

function BasketItem({ product }) {
  const dispatch = useDispatch();

  const { price, previousPrice } = usePrice({
    isOnSale: product.isOnSale,
    discount: product.discount * product.quantity,
    price: product.price * product.quantity,
    currencyCode: product.currencyCode,
  });

  const handleIncrementQuantity = () =>
    dispatch(incrementQuantity(product.sku));

  const handleDecrementQuantity = () =>
    dispatch(decrementQuantity(product.sku));

  const handleRemoveProduct = () => dispatch(removeProduct(product.sku));

  return (
    <Styled.ListItem>
      <Image
        src={product.imageUrl}
        alt=""
        width={200}
        height={200}
        quality={80}
        priority="true"
      />
      <div>
        <PartNumber fontSize={13}>{product.sku}</PartNumber>
        <Styled.ProductDescription>
          {product.name}, {product.description}
        </Styled.ProductDescription>
        <Styled.Container>
          <Price fontSize={26} fontWeight={500}>
            {price}
          </Price>
          <Styled.RemoveProduct type="button" onClick={handleRemoveProduct}>
            Remove Item
          </Styled.RemoveProduct>
        </Styled.Container>
        {product.isOnSale && (
          <PreviousPrice fontSize={15}>{previousPrice}</PreviousPrice>
        )}
        <StockDisplay stockStatus={product.stockStatus} marginTop="5px" />
        <Styled.QuantitySelector>
          <Styled.DecreaseButton
            type="button"
            title="Decrease Quantity"
            aria-label="Decrease Quantity"
            onClick={handleDecrementQuantity}
          >
            <Minus width={15} height={17} />
          </Styled.DecreaseButton>

          <Styled.Quantity data-testid="quantity">
            {product.quantity}
          </Styled.Quantity>

          <Styled.IncreaseButton
            type="button"
            title="Increase Quantity"
            aria-label="Increase Quantity"
            onClick={handleIncrementQuantity}
          >
            <Plus width={15} height={17} />
          </Styled.IncreaseButton>
        </Styled.QuantitySelector>
      </div>
    </Styled.ListItem>
  );
}

export default memo(BasketItem);
