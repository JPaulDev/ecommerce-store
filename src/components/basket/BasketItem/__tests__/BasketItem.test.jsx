import renderer from 'react-test-renderer';
import { screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../../../../app/store';
import { renderWithProviders } from '../../../../../test-utils';
import BasketItem from '../index';

// Prevents errors caused by next image component
jest.mock('next/future/image', () => 'img');

const product = {
  sku: 'CPU-INT-10400F',
  stock: 50,
  name: 'Intel Core i5 10400F',
  image: '',
  price: 129.99,
  salePrice: 114.99,
  quantity: 1,
  description: [
    '6 Cores, 12 Threads',
    'S1200 Comet Lake',
    '2.9GHz, 4.3GHz Turbo',
    '12MB Cache',
  ],
};

describe('BasketItem component', () => {
  it('renders a product card with an image and all product information', () => {
    const tree = renderer
      .create(
        <Provider store={setupStore()}>
          <BasketItem product={product} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('BasketItem component stock indicator', () => {
  it('product is displayed as in stock when 31 or more', () => {
    renderWithProviders(<BasketItem product={{ ...product, stock: 31 }} />);

    expect(screen.getByText(/in stock/i)).toBeInTheDocument();
  });

  it('product is displayed as low stock at 30 or fewer', () => {
    renderWithProviders(<BasketItem product={{ ...product, stock: 30 }} />);

    expect(screen.getByText(/low stock/i)).toBeInTheDocument();
  });

  it('product is displayed as out of stock at 0', () => {
    renderWithProviders(<BasketItem product={{ ...product, stock: 0 }} />);

    expect(screen.getByText(/out of stock/i)).toBeInTheDocument();
  });
});

describe('BasketItem component price display', () => {
  it('should display only the base price if no sale price is present', () => {
    renderWithProviders(
      <BasketItem product={{ ...product, salePrice: undefined }} />
    );

    expect(screen.getByText('129')).toHaveTextContent('£129.99');
    expect(screen.queryByText(/was/i)).not.toBeInTheDocument();
  });
});

describe('BasketItem component price totals', () => {
  it('should display the correct sale price total', () => {
    renderWithProviders(<BasketItem product={{ ...product, quantity: 2 }} />);

    expect(screen.getByText('229')).toHaveTextContent('£229.98');
  });

  it('should display the correct base price total', () => {
    renderWithProviders(<BasketItem product={{ ...product, quantity: 2 }} />);

    expect(screen.getByText('£259.98')).toBeInTheDocument();
  });
});
