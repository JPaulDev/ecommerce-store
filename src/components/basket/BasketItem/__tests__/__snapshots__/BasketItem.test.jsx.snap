// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasketItem component should render a product card with with all the product information 1`] = `
<li
  className="sc-13832a0-0 AhvBT"
>
  <img
    alt=""
    height={200}
    quality={85}
    src=""
    width={200}
  />
  <div>
    <span
      className="sc-bd55615-0 hWOyAS"
    >
      SKU: 
      GPU-3PAQ-ASU
    </span>
    <p
      className="sc-13832a0-1 hiFpEt"
    >
      ASUS ROG Strix GeForce RTX 3080 V2 OC
      , 
      1440MHz, 1905MHz Boost, 8704 CUDA Cores, 10GB GDDR6X
    </p>
    <div
      className="sc-13832a0-2 tqfkm"
    >
      <span
        className="sc-dded47fe-0 hFABLP"
      >
        <span
          className="sc-dded47fe-1 jIDwKn"
        >
          £
        </span>
        99
        <span
          className="sc-dded47fe-1 jIDwKn"
        >
          .99
        </span>
      </span>
      <button
        className="sc-13832a0-3 kcYllZ"
        onClick={[Function]}
        type="button"
      >
        Remove Item
      </button>
    </div>
    <span
      className="sc-2cd02fbb-0 dqciqJ"
    >
      <span
        className="sc-2cd02fbb-1 dlirfm"
      >
        Was: 
      </span>
      £129.99
    </span>
    <div
      className="sc-4ceb59c7-0 dHzdWC"
      title="Item is in stock"
    >
      <svg
        enableBackground="new 0 0 24 24"
        fill="var(--color-grey-8)"
        height={18}
        viewBox="0 0 24 24"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <rect
            fill="none"
            height="24"
            width="24"
          />
        </g>
        <g>
          <g>
            <path
              d="M5,5h2v3h10V5h2v5h2V5c0-1.1-0.9-2-2-2h-4.18C14.4,1.84,13.3,1,12,1S9.6,1.84,9.18,3H5C3.9,3,3,3.9,3,5v14 c0,1.1,0.9,2,2,2h6v-2H5V5z M12,3c0.55,0,1,0.45,1,1s-0.45,1-1,1s-1-0.45-1-1S11.45,3,12,3z"
            />
            <polygon
              points="21,11.5 15.51,17 12.5,14 11,15.5 15.51,20 22.5,13"
            />
          </g>
        </g>
      </svg>
      In Stock
    </div>
    <div
      className="sc-13832a0-4 esMEko"
    >
      <button
        aria-label="Decrease Quantity"
        className="sc-13832a0-5 sc-13832a0-6 gGkkzk cntabC"
        onClick={[Function]}
        title="Decrease Quantity"
        type="button"
      >
        <svg
          fill="currentColor"
          height={17}
          viewBox="0 0 448 512"
          width={15}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 256c0 17.7-14.3 32-32 32L48 288c-17.7 0-32-14.3-32-32s14.3-32 32-32l352 0c17.7 0 32 14.3 32 32z"
          />
        </svg>
      </button>
      <span
        className="sc-13832a0-7 hEMsnS"
        data-testid="quantity"
      >
        1
      </span>
      <button
        aria-label="Increase Quantity"
        className="sc-13832a0-5 gGkkzk"
        onClick={[Function]}
        title="Increase Quantity"
        type="button"
      >
        <svg
          fill="currentColor"
          height={17}
          viewBox="0 0 448 512"
          width={15}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
          />
        </svg>
      </button>
    </div>
  </div>
</li>
`;
